job.c

```diff
@@ -44,13 +44,13 @@ rehash()
     size_t old_cap = all_jobs_cap, old_used = all_jobs_used, i;
 
     if (cur_prime >= NUM_PRIMES) return;
-    if (hash_table_was_oom) return;
+    if (hash_table_max_prime <= cur_prime) return;
 
     all_jobs_cap = primes[++cur_prime];
     all_jobs = calloc(all_jobs_cap, sizeof(job));
     if (!all_jobs) {
         twarnx("Failed to allocate %zu new hash buckets", all_jobs_cap);
-        hash_table_was_oom = 1;
+        hash_table_max_prime = cur_prime;
```

Comment:
This won't work. A calloc failure here doesn't mean we've reached a maximum size for the hash table, it means the process *doesn't have enough memory* to allocate the new table. It's because of all the other objects (jobs, buffers, whatever) that are allocated at the moment, using up all that memory.

We could just clean up and return here, and then the next job added would immediately try to grow the hash table again, and likely fail, and same for the job after that, constantly trying (and failing) to allocate a new hash table, until a bunch of memory gets freed and then maybe it works again, or else jobs have been deleted so we don't actually need to grow the hash table any more.

That would be correct, but I had thought the call to calloc would be expensive, even if it fails, so as a simple hack to prevent that scenario of constantly failing to grow the hash table, I made the failure sticky. If it ever fails, it just gives up permanently and accepts whatever load factor fate bestows upon us.

That simple hack might not be good enough now that we're also trying to *shrink* the hash table. Presumably while deleting jobs we're also freeing memory, so if calloc fails in rehash_down, it might succeed the next time, or the time after that. I'd suggest starting simple though, and if we're shrinking, then just try every time, and not bother to check or set that oom flag. (But leave the existing behavior for the growing case.)

What do you think?